### ISSUE TITLE
Pass the repository to callbacks

### ISSUE THREAD

**Issue by josevalim** [elixir-lang/ecto#350]
We should either pass the repository explicitly or store them in the changeset.

Thoughts?

**Comment by briksoftware** [comment]
Is it a real use case that a model will be used in multiple repositories from the same app? Otherwise one could save the repo with a module attribute (say, @repo) on each model and this could be accessible for example with through the __schema__ function.

Since I doubt that having multiple repos in the same app at all is a very common use case, one could even define a @default_repo attribute somewhere, so you do not need to specify it on each model in the most common case.

After this is done, one can store the repo additionally in the changeset. So if someone develops an app with multiple repos he can explicitly set it/read it, but if they do not, the default one for the model or even for the app would always be easy accessible.

I do not know if my proposal is shortsighted, but having to pass the Repo around every time is a bit bothersome...

**Comment by josevalim** [comment]
The main reason is that you would be able to provide generic callbacks that work in any repo. For example, if the callback receives the repository, you can implement a callback that increments a counter cache column without asking the user to pass the repository explicitly:

    defmodule User do
      use Ecto.Model
      counter_cache :publications
    end

That's the main use case in my opinion.

> I do not know if my proposal is shortsighted, but having to pass the Repo around every time is a bit bothersome...

Note you woulnd't pass the callback around every time... you would simply receive it as argument. And, if we put it in the changeset, nothing would need to change! :)

**Comment by briksoftware** [comment]
Yes. My reasoning was, that if you need to perform any operation on data you need at least 2 pieces of information (let's say a `from` query, where I have no changeset yet):

1) the repo
2) the model (or model name)

If the repo were to be stored in the model, then you could do something like `MyModel.__schema__(:repo)` and only need to pass the model name.

In this case, storing it in the changeset would be redundant and the repo would be available as soon as you have a model, even if no changeset is available yet.

To better explain what I mean, let's take as example this code: https://github.com/briksoftware/crudex/blob/master/lib/crudex/crud_controller.ex

it provides a macro to generate a CRUD controller with Phoenix. It needs to know the module of the model for which it needs to be generated, but with the current approach in Ecto, it also needs the repo. With my proposal the repo would be readable from the model itself. However, the case where the same model is used by multiple repositories would be a "special case". The question is if this kind of flexibility (having a model with multiple repositories) is really needed.

Storing the repo in the changeset instead would allow more flexibility and would work perfectly for callbacks, but it would not cover the case I described above in my Crudex controller.

So, to summarize, I am not against storing the repo in the changeset (it is better than passing it explicitly imho) but I would consider the pro and contra of doing the Repo->Model association even before a changeset exists instead as well.

**Comment by josevalim** [comment]
> So, to summarize, I am not against storing the repo in the changeset (it is better than passing it explicitly imho) but I would consider the pro and contra of doing the Repo->Model association even before a changeset exists instead as well.

The jury is still out on this one.

The current thought is that it is better to couple those at the controller. So Crudex is doing the right thing by not pushing the coupling to the model.

It is also worth pointing out that it is not a discussion about how frequent or if multiple repositories are a common case. It is about better splitting the concerns instead of pushing all the responsibilities to the same piece of code.

For example, one consequence of the current design is that a master-slave setup or sharding is much, much easier to achieve, as you just need to implemented a proxy repo or a sharded based one. And that just came as a side-effect of not coupling. :)

**Comment by briksoftware** [comment]
I see. I certainly see the negative aspect of coupling repo with model and they have to be weighted against the pros. The more I think about it, the more I see you are actually right, even if having multiple repositories does not happen coupling is still something that could backfire in the long term.

The only significant pro would be a relatively lower entry barrier for people coming from other frameworks (Rails, to name one), since usually the concept of repository is hidden and all you see is the model. On the other hand if we try too hard to emulate ActiveRecords we will eventually end up simulating object orientation, which cannot be good in a functional language.

On the other hand, a changeset is a "thing" storing some transitional state. Storing a reference to the repo that caused the callback inside one does not create coupling in my opinion and makes it a little prettier than a separate explicit parameter for the callback.

**Comment by josevalim** [comment]
> On the other hand, a changeset is a "thing" storing some transitional state. Storing a reference to the repo that caused the callback inside one does not create coupling in my opinion and makes it a little prettier than a separate explicit parameter for the callback.

Agreed. Passing it as an argument, be it via the repo or the changeset, will help us avoid the coupling in the first place!
